Omdb-client

Garret-vin opened the pull request:
added omdb-client module with dto, api and implementation

Garret-vin committed:
added omdb-client module with dto, api and implementation

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -0,0 +1,56 @@
+    @Value("${api_url}")
+    private String url;
+    @Value("${api_key}")
+    private String apiKey;
```
Move it to PropertyPlaceholder and inject here as bean

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -0,0 +1,56 @@
+                .queryParam("r", "json")
+                .queryParam("s", title)
+                .queryParam("page", page);
+        while (hasNextPage) {
```
I don't think this is good place to do paging. Ideally batch job which you will add should do paging. So I think here you should retrurn something like OmdbResponse<List<ShortMovie>> which will also have page info

Dominys replied:
@Garret-vin Agree with @mmarusich Better to return page. If some service need to read more pages he will call this method multiple times.

Dominys commented on omdb-client/pom.xml:
```diff
@@ -0,0 +1,37 @@
+    <dependencies>
+        <dependency>
+            <groupId>com.garret.movies</groupId>
+            <artifactId>dao</artifactId>
```
@Garret-vin Why client need dao?? Especially api module.

Garret-vin replied:
@Dominys It was an outdated dependency. I removed it.

Dominys commented on omdb-client/pom.xml:
```diff
@@ -0,0 +1,37 @@
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-collections4</artifactId>
```
@Garret-vin All child modules need this library?

Garret-vin replied:
@Dominys It was outdated too

Dominys commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/config/OmdbConfig.java:
```diff
@@ -0,0 +1,34 @@
+public class OmdbConfig {
+
+    @Bean
+    public RestTemplate restTemplate() {
```
@Garret-vin use more meaningful name. for example "omdbClientRestTemplate"

Garret-vin committed:
fix comments

Garret-vin committed:
move string params to constructor

Garret-vin committed:
better solution then before for injecting string values as beans

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -3,19 +3,26 @@
     private final RestTemplate restTemplate;
 
+    public OmdbClientDTOImpl(@Value("${api_url}") String url,
+                             @Value("${api_key}") String apiKey,
```
https://www.baeldung.com/configuration-properties-in-spring-boot

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/config/OmdbConfig.java:
```diff
@@ -18,6 +19,16 @@
 public class OmdbConfig {
 
+    @Bean(name = "url")
+    public String omdbUrl(@Value("${api_url}") String url) {
```
Its not better solution :) previous was better

Garret-vin committed:
Merge branch 'master' into omdb_api_client

Garret-vin committed:
added configuration properties class+interface for api

Garret-vin committed:
Merge remote-tracking branch 'origin/omdb_api_client' into omdb_api_client

mmarusich commented on omdb-client/api/src/main/java/com/garret/movies/omdb/dto/MoviesResponse.java:
```diff
@@ -0,0 +1,16 @@
+@Data
+public class MoviesResponse {
+    @JsonProperty("Search")
+    private List<ShortMovie> shortMovieList;
```
What about making it generic?

mmarusich commented on omdb-client/api/src/main/java/com/garret/movies/omdb/dto/MoviesResponse.java:
```diff
@@ -0,0 +1,16 @@
+    @JsonProperty("Search")
+    private List<ShortMovie> shortMovieList;
+    @JsonProperty("totalResults")
+    private String totalResults;
```
I suppose this is number

mmarusich commented on omdb-client/api/src/main/java/com/garret/movies/omdb/dto/MoviesResponse.java:
```diff
@@ -0,0 +1,16 @@
+    @JsonProperty("totalResults")
+    private String totalResults;
+    @JsonProperty("Response")
+    private String response;
```
What value we get here?

Garret-vin replied:
@mmarusich true or false but in string type

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/config/TemplateConfig.java:
```diff
@@ -0,0 +1,38 @@
+public class TemplateConfig {
+
+    @Bean
+    public static PropertySourcesPlaceholderConfigurer propertySourcesPlaceholderConfigurer() {
```
why static? and why we need it? I think it should work out of the box

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -0,0 +1,36 @@
+        UriComponentsBuilder urlBuilder = UriComponentsBuilder.fromHttpUrl(clientConfig.getUrl())
+                .queryParam("apikey", clientConfig.getApiKey())
+                .queryParam("i", imdbId);
+        return restTemplate.getForObject(urlBuilder.toUriString(), OmdbMovie.class);
```
what it will return if no such movie exists anymore?

Garret-vin replied:
@mmarusich it will return MoviesResponse but in response var will be false

mmarusich replied:
But here you don't use MovesResponse, you use OmdbMovie

Garret-vin replied:
@mmarusich ups, i think about another method. This method return OmdbMovie, all right. I already added response var to OmdbMovie.

Garret-vin committed:
fix comments
added generic to omdbclient and MovieResponse
added bool var response to OmdbMovie
other small fixes

Garret-vin committed:
fix generic in responses
config json parsing for new responses

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -0,0 +1,54 @@
+                .queryParam("r", "json")
+                .queryParam("s", title)
+                .queryParam("page", page);
+        return restTemplate.exchange(
```
it almost same as in other method, extract it to some common method

mmarusich reviewed (approved):
LGTM

Garret-vin committed:
extract common logic into separate method

mmarusich commented on omdb-client/impl/src/main/java/com/garret/movies/omdb/impl/client/OmdbClientDTOImpl.java:
```diff
@@ -0,0 +1,54 @@
+        });
+    }
+
+    private <T> T sendGetRequestToApi(UriComponentsBuilder urlBuilder, ParameterizedTypeReference<T> typeReference) {
```
don't bass builder, you just get url here, pass ready url

Garret-vin committed:
pass string url instead builder
